src/plugins/socketmon/socketmon.cpp

```diff
@@ -1379,6 +1382,287 @@ static event_response_t udpb_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
 
 /* ----------------------------------------------------- */
 
+static void print_dns_info(drakvuf_t drakvuf, drakvuf_trap_info_t* info, socketmon* sm, const char* function_name, const char* dns_name)
+{
+    switch (sm->format)
+    {
+        case OUTPUT_CSV:
+            printf("socketmon," FORMAT_TIMEVAL ",%" PRIu32 ",0x%" PRIx64 ",\"%s\",\"%s\",\"%s\"\n",
+                   UNPACK_TIMEVAL(info->timestamp), info->vcpu, info->regs->cr3,
+                   info->proc_data.name, dns_name, function_name);
+            break;
+
+        case OUTPUT_KV:
+            printf("socketmon Time=" FORMAT_TIMEVAL ",PID=%d,PPID=%d,ProcessName=\"%s\",DnsName=\"%s\",Method=\"%s\"\n",
+                   UNPACK_TIMEVAL(info->timestamp), info->proc_data.pid, info->proc_data.ppid,
+                   info->proc_data.name, dns_name, function_name);
+            break;
+
+        default:
+        case OUTPUT_DEFAULT:
+            printf("[SOCKETMON] TIME:" FORMAT_TIMEVAL " VCPU:%" PRIu32 " CR3:0x%" PRIx64 ",\"%s\" DNS_NAME:\"%s\" METHOD:\"%s\"\n",
+                   UNPACK_TIMEVAL(info->timestamp), info->vcpu, info->regs->cr3, info->proc_data.name,
+                   dns_name, function_name);
+            break;
+    };
+}
+
+static event_response_t trap_DnsQuery_A_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    socketmon* sm = (socketmon*)info->trap->data;
+
+    addr_t domain_name_addr = 0;
+
+    if (sm->pm == VMI_PM_IA32E) // only 64 bit for now
+    {
+        domain_name_addr = info->regs->rcx;
+
+        access_context_t ctx;
+        ctx.translate_mechanism = VMI_TM_PROCESS_DTB;
+        ctx.dtb = info->regs->cr3;
+        ctx.addr = domain_name_addr;
+
+        char* dns_name = [&]
```

Comment:
Huh, I'm really not familiar with this syntax. What does this do and could it changed to something that's more standard?

Reply:
This is a lambda function from C++11, it creates a function. 
Here this function that is instantly called after creation. 

[&] means that it captures all variables by reference.

Here is more info on lambdas:
https://en.cppreference.com/w/cpp/language/lambda

Reply:
That is pretty cool but my worry about advanced C++ features like this is they tend to extremely slow. Even constructors tend to be extremely slow. So using these functions in event callbacks is problematic.

Reply:
I did some experimentation with doing the lambda wrapper around the code here and it made no visible difference performance-wise so I'm good to move ahead with it.

Reply:
Thank you!